File: typescript/src/2021/day-05.ts
import { Solution } from "../solution";
import { HashMap } from "../collections";
import { print, range } from "../utils";

class Point {
  x: number;
  y: number;

  constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }

  toString() {
    return `(${this.x},${this.y})`;
  }
}

class Line {
  start: Point;
  end: Point;

  constructor(start: Point, end: Point) {
    this.start = start;
    this.end = end;
  }

  toString() {
    return `${this.start} => ${this.end}`;
  }

  isHorizontal() {
    return this.start.y === this.end.y;
  }

  isVertical() {
    return this.start.x === this.end.x;
  }
}

class Day5 extends Solution {
  constructor(example?: number) {
    super(5, 2021, example);
  }

  part1(): number | string {
    const lines = this.readInput().map(this.parseLine);
    const hvLines = lines.filter(l => l.isHorizontal() || l.isVertical());

    const overlaps = new HashMap<Point, number>();

    hvLines.forEach(line => {
      if (line.isHorizontal()) {
        const minX = Math.min(line.start.x, line.end.x);
        const maxX = Math.max(line.start.x, line.end.x);
        for (let i = minX; i <= maxX; i++) {
          const point = new Point(i, line.start.y);
          const numOverlaps = overlaps.get(point) || 0;
          overlaps.set(point, numOverlaps + 1);
        }
      } else if (line.isVertical()) {
        const minY = Math.min(line.start.y, line.end.y);
        const maxY = Math.max(line.start.y, line.end.y);
        for (let i = minY; i <= maxY; i++) {
          const point = new Point(line.start.x, i);
          const numOverlaps = overlaps.get(point) || 0;
          overlaps.set(point, numOverlaps + 1);
        }
      }
      // print(line);
      // print(overlaps);
    });

    return overlaps.values().filter(v => v > 1).length;
  }

  parseLine(line: string): Line {
    const splits = line.split(" ");
    const [x, y] = splits[0].split(",").map(n => parseInt(n, 10));
    const [x2, y2] = splits[2].split(",").map(n => parseInt(n, 10));
    return new Line(new Point(x, y), new Point(x2, y2));
  }

  part2(): number | string {
    const lines = this.readInput().map(this.parseLine);

    const overlaps = new HashMap<Point, number>();

    lines.forEach(line => {
      if (line.isHorizontal()) {
        for (const x of range(line.end.x, line.start.x, true)) {
          const point = new Point(x, line.start.y);
          const numOverlaps = overlaps.get(point) || 0;
          overlaps.set(point, numOverlaps + 1);
        }
      } else if (line.isVertical()) {
        for (const y of range(line.end.y, line.start.y, true)) {
          const point = new Point(line.start.x, y);
          const numOverlaps = overlaps.get(point) || 0;
          overlaps.set(point, numOverlaps + 1);
        }
      } else {
        const xs = range(line.end.x, line.start.x, true);
        const ys = range(line.end.y, line.start.y, true);
        for (let i = 0; i < xs.length; i++) {
          const point = new Point(xs[i], ys[i]);
          const numOverlaps = overlaps.get(point) || 0;
          overlaps.set(point, numOverlaps + 1);
        }
      }
      // print(line);
      // print(overlaps);
    });

    // this.printGrid(overlaps);

    return overlaps.values().filter(v => v > 1).length;
  }

  printGrid(overlaps: HashMap<Point, number>) {
    let g = "";
    const maxX = Math.max(...overlaps.keys().map(k => k.x));
    const maxY = Math.max(...overlaps.keys().map(k => k.y));

    for (let y = 0; y <= maxY; y++) {
      for (let x = 0; x <= maxX; x++) {
        const p = new Point(x, y);
        const o = overlaps.get(p) || 0;
        if (o === 0) {
          g += ".";
        } else {
          g += o.toString();
        }
      }
      g += "\n";
    }

    print(g);
  }
}

(new Day5()).run();

File: java/src/main/java/com/weizilla/adventofcode/y2023/Day05.java
package com.weizilla.adventofcode.y2023;

import com.google.common.collect.Range;
import com.weizilla.adventofcode.utils.Day;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.NavigableMap;
import java.util.TreeMap;

public class Day05 extends Day {
    public Day05(Integer example) {
        super(example);
    }

    @Override
    public Object part1() {
        var lines = reader.readStrings();

        var startSeeds = Arrays.stream(lines.get(0).split(":")[1].trim().split(" "))
            .map(Long::valueOf)
            .toList();

//        print("{}", startSeeds);

        var catMaps = parseMaps(lines);
//        print("{}", catMaps);

        Cat currCat = Cat.SEED;
        var currValues = new ArrayList<Long>(startSeeds);
        while (currCat != Cat.LOCATION) {
            var catMap = catMaps.get(currCat);

            var nextValues = new ArrayList<Long>();
            for (long currValue : currValues) {
                nextValues.add(catMap.getDestination(currValue));
            }
            currValues = nextValues;
            currCat = currCat.getDestination();
        }

        return currValues.stream().min(Long::compareTo).get();
    }

    private Map<Cat, CatMap> parseMaps(List<String> lines) {

        var maps = new HashMap<Cat, CatMap>();
        CatMap curr = null;
        for (int i = 1; i < lines.size(); i++) {
            String line = lines.get(i);
            if (line.isEmpty()) {
                continue;
            }

            if (line.contains("map")) {
                curr = new CatMap();
                var type = Cat.valueOf(line.split("-")[0].toUpperCase());
                maps.put(type, curr);
            } else {
                var splits = line.split(" ");
                curr.addRange(
                    Long.parseLong(splits[0]),
                    Long.parseLong(splits[1]),
                    Integer.parseInt(splits[2])
                );
            }
        }

        return maps;
    }

    @Override
    public Object part2() {
        var lines = reader.readStrings();

        var startValues = Arrays.stream(lines.get(0).split(":")[1].trim().split(" "))
            .map(Long::valueOf)
            .toList();

        List<Range<Long>> currValues = new ArrayList<>();
        for (int i = 0; i < startValues.size(); i += 2) {
            currValues.add(Range.closed(startValues.get(i), startValues.get(i) + startValues.get(i + 1)));
        }

        print("{}", currValues);

        var catMaps = parseMaps(lines);
//        print("{}", catMaps);

        Cat currCat = Cat.SEED;
        while (currCat != Cat.LOCATION) {
            var catMap = catMaps.get(currCat);
            currValues = map(currValues, catMap);
            currCat = currCat.getDestination();
        }

        return currValues.stream().mapToLong(Range::lowerEndpoint).min();
    }


    private List<Range<Long>> map(List<Range<Long>> values, CatMap mapping) {
        List<Range<Long>> result = new ArrayList<>();

        values.stream()
            .map(e -> map(e, mapping))
            .forEach(result::addAll);

        return result;
    }

    private List<Range<Long>> map(Range<Long> ranges, CatMap allMappings) {
        List<Range<Long>> unmapped = List.of(ranges);
        List<Range<Long>> mapped = new ArrayList<>();

        for (Mapping mapping : allMappings.getMappings().values()) {

            List<Range<Long>> newUnmapped = new ArrayList<>();
            for (Range<Long> curr : unmapped) {
                if (curr.upperEndpoint() < mapping.source() || curr.lowerEndpoint() > mapping.sourceEnd()) {
                    newUnmapped.add(curr);
                    continue;
                }

                long centerLowerMapped = 0;
                long centerUpperMapped = 0;

                if (curr.lowerEndpoint() < mapping.source()) {
                    // ----
                    //  curr  ----
                    //        mapping
                    Range<Long> lowerDiff = Range.closed(curr.lowerEndpoint(), mapping.source());
                    newUnmapped.add(lowerDiff);
                    centerLowerMapped = mapping.destination();
                } else if (curr.lowerEndpoint() >= mapping.source()) {
                    //        ----
                    // ----   mapping
                    //  curr
                    centerLowerMapped = curr.lowerEndpoint() - mapping.source() + mapping.destination();
                }

                if (curr.upperEndpoint() <= mapping.sourceEnd()) {
                    // curr
                    // ----  mapping
                    //       ----

                    centerUpperMapped = mapping.destEnd() - (mapping.sourceEnd() - curr.upperEndpoint());
                } else if (curr.upperEndpoint() > mapping.sourceEnd()) {
                    //        mapping
                    // curr   ----
                    // ----
                    centerUpperMapped = mapping.destEnd();
                    Range<Long> upperDiff = Range.closed(mapping.sourceEnd(), curr.upperEndpoint());
                    newUnmapped.add(upperDiff);
                }

                mapped.add(Range.closed(centerLowerMapped, centerUpperMapped));
            }
            unmapped = newUnmapped;
        }

        mapped.addAll(unmapped);


        return mapped;
    }

    private enum Cat {
        LOCATION(null),
        HUMIDITY(LOCATION),
        TEMPERATURE(HUMIDITY),
        LIGHT(TEMPERATURE),
        WATER(LIGHT),
        FERTILIZER(WATER),
        SOIL(FERTILIZER),
        SEED(SOIL);

        private final Cat destination;

        Cat(Cat destionation) {
            this.destination = destionation;
        }

        public Cat getDestination() {
            return destination;
        }
    }

    private static class CatMap {
        private final NavigableMap<Long, Mapping> mappings;

        public CatMap() {
            mappings = new TreeMap<>();
        }

        public void addRange(long destination, long source, long length) {
            mappings.put(source, new Mapping(source, destination, length));
        }

        public long getDestination(long source) {
            var entry = mappings.floorEntry(source);
            if (entry == null) {
                return source;
            }
            var start = entry.getKey();
            var length = entry.getValue().length();
            var destination = entry.getValue().destination();
            if (start + length < source) {
                return source;
            }
            return destination + source - start;
        }

        public NavigableMap<Long, Mapping> getMappings() {
            return mappings;
        }


        @Override
        public String toString() {
            return "CatMap{" +
                "mappings=" + mappings +
                '}';
        }
    }

    private record Mapping(long source, long destination, long length) {
        public long sourceEnd() {
            return source + length;
        }

        public long destEnd() {
            return destination + length;
        }
    }

}
